parts/k8s/kubernetescustomscript.sh

```diff
@@ -62,7 +62,7 @@ else
 fi
 
 function testOutboundConnection() {
-    retrycmd_if_failure 20 1 3 nc -v 8.8.8.8 53 || retrycmd_if_failure 20 1 3 nc -v 8.8.4.4 53 || exit $ERR_OUTBOUND_CONN_FAIL
+    retrycmd_if_failure 20 1 3 nc -v www.google.com 443 || retrycmd_if_failure 20 1 3 nc -v www.1688.com 443 || exit $ERR_OUTBOUND_CONN_FAIL
```

Comment:
Noticed that 8.8.8.8 and 8.8.8.4 are listening on TCP 443, so let's make one final change and test the IP directly to avoid the DNS lookup (we're not testing for DNS resolution here).

Reply:
Hey @jackfrancis from my point of view -- testing for name resolution is essential at some point early in the process, we have found customer sites (private Azure Cloud) which have custom DNS that was misconfigured in that ARR is required for name resolution within the cluster, and the customer AAD was required for their intranet which we inter-operate with.

What's the downside of performing name resolution at this point in the startup process?

Reply:
also, I have not yet been able to find an official google doc that port 443 has formal long term support at that IP address

Reply:
> What's the downside of performing name resolution at this point in the startup process?

I think the issue is that this not only changes the port from 53 to the safer 443, but that it switches from IPs to hostnames. The DNS lookup arguably wasn't intended to be part of `testOutboundConnection`.

In general, it's probably just as reliable to contact `www.google.com:443` as `8.8.8.8:53`, and I agree with your point that `8.8.8.8:443` might not be officially supported. We were just wondering if there's a way to do this that doesn't make the test (slightly) more complicated by using a hostname--maybe there isn't.

Reply:
I guess this is why they are listening on 443 https://developers.google.com/speed/public-dns/docs/dns-over-https. but this still is not a best option to use, imho. google dns IPs may be blocked completely inside of an org network. I just tested 2 different org networks I have access to, and it works in one but does not work in another one. May I ask why do you need to test "internet connectivity" this way? I'd better try accessing MSFT endpoints, for instance, like login.microsoft.com or portal.azure.com over 443, or something like that. But definitely not DNS, as this service may be considered "insecure" inside of an org

Reply:
@jackfrancis @CecileRobertMichon please also keep in mind that there could be some kind of "proxy" related issues, when you may need to authenticate against proxy or kind of transparent proxy, before it will be able to forward your http(s) connection to somewhere. So sending simple TCP-SYN may not work.

Reply:
@jackfrancis got it. One thing, it maybe off topic but still. I'd think of a possibility to first download all required binaries into a storage count, aside of cluster provisioning. this step can be done by an admin or something. For instance I run a script on my laptop which downloads all needed binaries, and then I can upload them manually to a storage account. And then I just point acs-engine to that storage account to pick binaries from it. In this case we may not even need an internet connectivity. Service Endpoint will do the trick.

Reply:
@eosfor We're actually in the process of baking all of these dependencies onto an acs-engine-specific disk image, ETA for that is 4-8 weeks.

Reply:
@jackfrancis just came into my mind, there is a 168.63.129.16 address. Maybe you in MSFT can leverage it to get to the AKS related binaries. This one is required to be accessible for all the stuff in Azure, so, i'd say it is better choice.

Reply:
We're actually hard at work getting pre-baked OS images over the finish line, after which we'll remove this CSE check entirely (we will no longer need to call out to the internet to scaffold a k8s node).

Reply:
@jackfrancis, will you announce this somehow? Will it affect AKS?